std/range/primitives.d

```diff
@@ -123,16 +123,26 @@ r.popFront();     // can invoke popFront()
 auto h = r.front; // can get the front of the range of non-void type
 ----
 
-The semantics of an input range (not checkable during compilation) are
-assumed to be the following ($(D r) is an object of type $(D R)):
-
-$(UL $(LI $(D r.empty) returns $(D false) iff there is more data
-available in the range.)  $(LI $(D r.front) returns the current
-element in the range. It may return by value or by reference. Calling
-$(D r.front) is allowed only if calling $(D r.empty) has, or would
-have, returned $(D false).) $(LI $(D r.popFront) advances to the next
-element in the range. Calling $(D r.popFront) is allowed only if
-calling $(D r.empty) has, or would have, returned $(D false).))
+The following are rules of input ranges are assumed to hold true in all
+Phobos code. These rules are not checkable at compile-time, so not conforming
+to these rules when writing ranges or range based code will result in
+undefined behavior.
+
+$(UL
+    $(LI `r.empty` returns `false` iff there is more data
+    available in the range.)
+    $(LI `r.empty` called multiple times, without calling
+    `r.popFront`, yields the same result for every call.)
+    $(LI `r.front` returns the current element in the range.
+    It may return by value or by reference.)
+    $(LI `r.front` can be legally called iff calling
+    `r.empty` has, or would have, returned `false`.)
+    $(LI `r.front` called multiple times, without calling
+    `r.popFront`, yields the same result for every call.)
+    $(LI `r.popFront` advances to the next element in the range.)
+    $(LI `r.popFront` can be called iff calling `r.empty`
+    has, or would have, returned `false`.)
```

Comment:
Again for the sake of pedantry: "User code may call r.popFront without having called r.empty before it if it has other means to ascertain that the range cannot be empty."